### PR TITLE
osd, crimson/osd: do_cmp_xattr() related cleanups

### PR DESCRIPTION
**Pull request by tchaikov**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/pg_backend.cc [hunk 1]
```diff
@@ -830,30 +830,32 @@ PGBackend::get_attr_ierrorator::future<> PGBackend::get_xattrs(
...
   });
 }
-static int do_xattr_cmp_str(int op, const string& lhs, bufferlist& rhs_xattr)
-{
-  string rhs(rhs_xattr.c_str(), rhs_xattr.length());
 
-  logger().debug("do_xattr_cmp_str '{}' vs '{}' op {}", lhs, rhs, op);
+namespace {
 
+template<typename U, typename V>
```
Comment (rzarzynski) [inline src/crimson/osd/pg_backend.cc:837]: ~~Do we need `V`? Is there a case where it differs from `U`?~~

`string` and `string_view` in `PGBackend::cmp_xattr`.

File: src/crimson/osd/pg_backend.cc [hunk 2]
```diff
@@ -900,10 +887,11 @@ PGBackend::cmp_xattr_ierrorator::future<> PGBackend::cmp_xattr(
     switch (osd_op.op.xattr.cmp_mode) {
     case CEPH_OSD_CMPXATTR_MODE_STRING:
       {
-        string val;
-        bp.copy(osd_op.op.xattr.value_len, val);
-        result = do_xattr_cmp_str(osd_op.op.xattr.cmp_op, val, xattr);
-        logger().debug("cmpxattr val={}, xattr= {}", val, xattr);
+        string lhs;
+        bp.copy(osd_op.op.xattr.value_len, lhs);
```
Comment (rzarzynski) [inline src/crimson/osd/pg_backend.cc:896]: `lhs` is a temporary; we don't to extend the life-time. This makes me think about a variant of `copy` that would fill `string_view`.

**Review by rzarzynski — APPROVED** [review]

File: src/crimson/osd/pg_backend.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/osd/pg_backend.cc:896]: as long as we can ensure that the bufferlist is contiguous. otherwise we pay for the overhead of rebuilding it.

**Review by liu-chunmei — APPROVED** [review]
LGTM.

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-06-29_13:42:32-rados-wip-kefu-testing2-2021-06-29-1946-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-30_04:57:30-rados-wip-kefu-testing2-2021-06-30-1019-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-06-30_08:47:33-rados-wip-kefu-testing2-2021-06-30-1405-distro-basic-smithi/

**Pull request merged**